lib/src/Pos/Block/Logic/Header.hs

```diff
@@ -353,34 +353,67 @@ getHeadersOlderExp upto = do
                 | otherwise = selGo es ii $ succ skipped
         in selGo elems ixs 0
 
--- CSL-396 don't load all the blocks into memory at once
 -- | Given @from@ and @to@ headers where @from@ is older (not strict)
 -- than @to@, and valid chain in between can be found, headers in
 -- range @[from..to]@ will be found.
-getHeadersFromToIncl
+getHeadersRange
     :: forall m. (HasConfiguration, MonadDBRead m)
-    => HeaderHash -> HeaderHash -> m (Maybe (OldestFirst NE HeaderHash))
-getHeadersFromToIncl older newer = runMaybeT . fmap OldestFirst $ do
+    => HeaderHash -> HeaderHash -> m (Either Text (OldestFirst NE HeaderHash))
+getHeadersRange older newer | older == newer = runExceptT $ do
+    unlessM (isJust <$> DB.getHeader newer) $
+        throwError "getHeadersRange: can't find newer-older header"
+    pure $ OldestFirst $ one newer
+getHeadersRange older newer = runExceptT $ do
     -- oldest and newest blocks do exist
-    start <- MaybeT $ DB.getHeader newer
-    end   <- MaybeT $ DB.getHeader older
-    guard $ getEpochOrSlot start >= getEpochOrSlot end
-    let lowerBound = getEpochOrSlot end
-    if newer == older
-    then pure $ one newer
-    else loadHeadersDo lowerBound (one newer) $ start ^. prevBlockL
+    newerHd <- fromMaybeM "can't retrieve newer header" $ DB.getHeader newer
+    olderHd <- fromMaybeM "can't retrieve older header" $ DB.getHeader older
+    let olderD = olderHd ^. difficultyL
+    let newerD = newerHd ^. difficultyL
+
+    -- Proving newerD >= olderD
+    when (newerD == olderD) $
+        throwError $ "getHeadersRange: newer and older headers have " <>
+                     "the same difficulty, but are not equal"
+    when (newerD < olderD) $
+        throwError "getHeadersRange: newer header is less dificult that older one"
+
+    -- How many epochs does this range cross.
+    let genDiff :: Int
+        genDiff = fromIntegral $ newerHd ^. epochIndexL - olderHd ^. epochIndexL
+    -- Number of blocks is difficulty difference + number of genesis blocks.
+    let depthDiff :: Int
+        depthDiff = genDiff + fromIntegral (newerD - olderD)
+
+    -- We load these depthDiff blocks.
+    let cond _curHash depth = depth < depthDiff
+
+    -- This is [oldest..newest) headers.
+    allExceptNewest <- GS.loadHashesUpWhile older cond
+
+    -- Sometimes we will get an empty list, if we've just switched the
+    -- branch (after first checks are performed here) and olderHd is
+    -- no longer in the main chain.
+    when (null $ allExceptNewest ^. _OldestFirst) $ throwError $
```

Comment:
We can use snapshots to avoid it, btw.

Reply:
Indeed. It's not really important here, but if you would like to see it, please tell me, I'll figure it out.

Reply:
Can you create an issue and let's postpone it (~for eternity~)?

Reply:
https://iohk.myjetbrains.com/youtrack/issue/CSL-1950